specification/dartLangSpec.tex

```diff
@@ -9685,29 +9691,113 @@ \subsubsection{Cascaded Invocations}
   \alt <selector>* (<assignableSelector> <cascadeAssignment>)?
 
 <cascadeAssignment> ::= <assignmentOperator> <expressionWithoutCascade>
-
-<argumentPart> ::=
-  <typeArguments>? <arguments>
 \end{grammar}
 
 \LMHash{}%
-Evaluation of a cascaded method invocation expression $c$ of the form \code{$e$..\metavar{suffix}} proceeds as follows:
+A \IndexCustom{cascaded member access}{cascaded member access}
+is an expression of the form
+\code{$e$..$s$} or \code{$e$?..$s$},
+where $e$ is an expression
+and $s$ is derived from \synt{cascadeSection}.
+The form
+\code{$e$?..$s$}
+is also known as a
+\IndexCustom{conditional cascaded member access}{%
+  cascaded member access!conditional}.
 
-Evaluate $e$ to an object $o$.
-Let $t$ be a fresh variable bound to $o$.
-Evaluate \code{$t$.\metavar{suffix}} to an object.
-Then $c$ evaluates to $o$.
+\rationale{%
+A \synt{cascadeSection} allows for accessing members, including setters.
+The motivation for having a cascaded member access is that it allows for
+performing a chain of operations based on an object
+while preserving a reference to that object for further processing.%
+}
+
+\commentary{%
+For example, \code{C()..foo.bar = 2} allows us to obtain a reference to
+the object $o$ which is the result of evaluating \code{C()},
+and at the same time use $o$ to invoke the getter \code{foo} and
+the setter \code{bar=} on the value returned by that getter.%
+}
 
 \LMHash{}%
-The static type of $c$ is the static type of $e$.
+An expression of the form \code{$e_0$?..$s$} where
+$e_0$ is a \synt{conditionalExpression} and
+$s$ is a \synt{cascadeSection} is an
+\IndexCustom{initially conditional cascaded member access}{%
+  cascaded member access!initially conditional}.
+Moreover, if \code{$e_0$} is
+an initially conditional cascaded member access and
+$s$ is derived from \synt{cascadeSection} then
+\code{$e_0$..$s$} is also initially conditional.
 
-\rationale{
-With the introduction of null-aware conditional assignable expressions (\ref{assignableExpressions}), it would make sense to extend cascades with a null-aware conditional form as well.
-One might define \code{$e$?..\metavar{suffix}} to be equivalent to the expression \code{$t$ == \NULL{} ? \NULL{} : $t$.\metavar{suffix}} where $t$ is a fresh variable bound to the value of $e$.
+\commentary{%
+In short, a cascade is initially conditional if
+the ``innermost dots'' are \lit{?..} rather than \lit{..}.
+Note that only the innermost dots can have the \lit{?}.
+All the non-innermost ones are implicitly skipped if the receiver is null,
+so any \lit{?} on a non-innermost \lit{..} would be useless.%
+}
+
+\LMHash{}%
+Let $s$ be a \synt{cascadeSection}.
+Let $e$ be an expression of the form \code{$e_0$..$s$}
+where $e_0$ is an \synt{expression}
+such that $e$ is not initially conditional.
+Exactly the same compile-time errors are raised for $e$
+as would be raised for \code{$e_0$.$s$}.
```

Comment:
Make that  ($e_0$).$s$  because otherwise e0 == x..foo would mean that `x..foo..bar` is the same as `x..foo.bar`.

Reply:
Did that already. It is also needed for other reasons: `(a + b).x` etc. may be ugly, but should still be handled correctly. ;-)